### PR TITLE
feat: support the attributeDescriptor X.509v3 extension

### PR DESCRIPTION
**Pull request by JonathanWilbur**
Support the  `attributeDescriptor` X.509v3 extension as defined in [ITU-T Recommendation X.509](https://www.itu.int/rec/T-REC-X.509/en). The syntax of this extension is:

```asn
attributeDescriptor EXTENSION ::= {
  SYNTAX         AttributeDescriptorSyntax
  IDENTIFIED BY  {id-ce-attributeDescriptor} }

AttributeDescriptorSyntax ::= SEQUENCE {
  identifier             AttributeIdentifier,
  attributeSyntax        OCTET STRING(SIZE (1..MAX)),
  name              [0]  AttributeName OPTIONAL,
  description       [1]  AttributeDescription OPTIONAL,
  dominationRule         PrivilegePolicyIdentifier,
  ... }

AttributeIdentifier ::= ATTRIBUTE.&id({AttributeIDs})

AttributeIDs ATTRIBUTE ::= {...}

AttributeName ::= UTF8String(SIZE (1..MAX))

AttributeDescription ::= UTF8String(SIZE (1..MAX))

PrivilegePolicyIdentifier ::= SEQUENCE {
  privilegePolicy  PrivilegePolicy,
  privPolSyntax    InfoSyntax,
  ... }

PrivilegePolicy ::= OBJECT IDENTIFIER

InfoSyntax ::= CHOICE {
  content  UnboundedDirectoryString,
  pointer  SEQUENCE {
    name     GeneralNames,
    hash     HASH{HashedPolicyInfo} OPTIONAL,
    ... },
  ... }

POLICY ::= TYPE-IDENTIFIER

HashedPolicyInfo ::= POLICY.&Type({Policies})
```

##### Checklist
- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Review by t8m — APPROVED** [review]

**Comment by JonathanWilbur** [comment]
@t8m I fixed the merge conflicts introduced from merging in the `roleSpecCertIdentifier` X.509v3 extension, so this is ready for review again. @nhorman would you be interested in joining?

**Comment by nhorman** [comment]
I will make an effort, but I'm away from my desk for a few days

File: crypto/x509/v3_attrdesc.c [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    }
+    return 1;
+}
+
+static int i2r_OSSL_INFO_SYNTAX(X509V3_EXT_METHOD *method,
+                                OSSL_INFO_SYNTAX *info,
+                                BIO *out, int indent)
+{
+    switch (info->type) {
+    case (OSSL_INFO_SYNTAX_TYPE_CONTENT):
```
Comment (FdaSilvaYY) [inline crypto/x509/v3_attrdesc.c:97]: nit: useless parenthesis

File: crypto/x509/v3_attrdesc.c [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        if (BIO_printf(out, "%*sContent: ", indent, "") <= 0)
+            return 0;
+        if (BIO_printf(out, "%.*s", info->choice.content->length, info->choice.content->data) <= 0)
+            return 0;
+        if (BIO_puts(out, "\n") <= 0)
+            return 0;
+        return 1;
+    case (OSSL_INFO_SYNTAX_TYPE_POINTER):
```
Comment (FdaSilvaYY) [inline crypto/x509/v3_attrdesc.c:105]: nit; parenthesis

File: crypto/x509/v3_utl.c [hunk 3]
```diff
@@ -1434,3 +1434,20 @@ int OSSL_GENERAL_NAMES_print(BIO *out, GENERAL_NAMES *gens, int indent)
     }
     return 1;
 }
+
+int ossl_bio_print_hex(BIO *out, unsigned char *buf, int len)
+{
+    int result = 1;
```
Comment (FdaSilvaYY) [inline crypto/x509/v3_utl.c:1440]: useless init.

**Comment by JonathanWilbur** [comment]
All feedback addresses. This PR is ready for review again.

**Comment by JonathanWilbur** [comment]
@t8m and @nhorman can I get a review on this?

**Review by t8m — APPROVED** [review]

File: crypto/x509/v3_attrdesc.c [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        return 0;
+    if (BIO_printf(out, "%*sSyntax:\n", indent, "") <= 0)
+        return 0;
+    return i2r_OSSL_INFO_SYNTAX(method, ppid->privPolSyntax, out, indent + 4);
+}
+
+static int i2r_OSSL_ATTRIBUTE_DESCRIPTOR(X509V3_EXT_METHOD *method,
+                                         OSSL_ATTRIBUTE_DESCRIPTOR *ad,
+                                         BIO *out, int indent) {
+    char buf[80];
```
Comment (nhorman) [inline crypto/x509/v3_attrdesc.c:135]: are we guaranteed to have an object identifier less than 80 chars?  Might it be better to use a BIO_s_mem here?

**Review by nhorman — COMMENTED** [review]
just the one nit I can find so far

File: crypto/x509/v3_attrdesc.c (hunk 4 above)
Comment (JonathanWilbur) [reply crypto/x509/v3_attrdesc.c:135]: I just copied the 80-bytes standard from here: https://github.com/openssl/openssl/blob/50e9d2b188b8dce070f388640c06a7dc04417390/crypto/asn1/tasn_prn.c#L404-L415

But I would be happy to do BIO_s_mem instead.

Comment (JonathanWilbur) [reply crypto/x509/v3_attrdesc.c:135]: @nhorman or perhaps @t8m do you have an opinion on the above?

Comment (FdaSilvaYY) [reply crypto/x509/v3_attrdesc.c:135]: @nhorman if you grep "[80]" in the codebase, you will find that it is a widespread practice for query SN/LN names from OIDs.

**Review by FdaSilvaYY — APPROVED** [review]

**Comment by FdaSilvaYY** [comment]
Could you please check command output to avoid repeating #25814 ?

File: crypto/x509/v3_attrdesc.c (hunk 4 above)
Comment (JonathanWilbur) [reply crypto/x509/v3_attrdesc.c:135]: @nhorman can I mark this as resolved?

Comment (nhorman) [reply crypto/x509/v3_attrdesc.c:135]: If 80 chars is the de-facto standard in our code, then I'm fine with it, though I'd be curious to know if thats codified in a standard anywhere.  Thats a future question though, we can resolve this

**Comment by JonathanWilbur** [comment]
Here is the entire content of the output of `./apps/openssl x509 -in ./test/certs/ext-attributeDescriptor.pem -noout -text` with this PR. @FdaSilvaYY hopefully this assuages your concern. I do visually check my PRs when I run them!

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 66051 (0x10203)
        Signature Algorithm: sha1WithRSAEncryption
        Issuer: 
        Validity
            Not Before: Aug 31 01:07:09 2021 GMT
            Not After : Aug 31 01:07:09 2021 GMT
        Subject: 
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:b6:78:cb:9b:5b:6c:d6:10:b8:7c:d3:6d:dd:49:
                    d0:0f:dc:bb:dd:b0:d7:82:2a:13:c9:66:12:23:77:
                    1a:fc:a3:5f:b9:3c:9d:8c:54:c4:60:09:aa:9e:e1:
                    ab:04:e5:1f:81:b9:87:dc:f4:50:d0:09:a9:aa:58:
                    e0:59:36:f7:77:72:bc:1f:85:05:bc:35:90:4c:f4:
                    92:db:52:0c:8e:6f:28:a9:dd:7d:9c:4e:46:c5:68:
                    ab:1a:ed:28:0d:1c:e1:58:b1:de:d5:16:7b:66:ba:
                    21:08:f8:3f:d4:e7:2c:cb:8e:d0:66:e2:b6:95:a1:
                    41:d2:b1:26:ad:15:81:98:46:1b:0e:5d:ff:c3:17:
                    f9:5d:fa:88:aa:93:b2:9c:90:74:da:d8:57:ad:30:
                    9c:4c:ce:d1:cf:51:6a:08:5b:7f:65:56:41:e6:12:
                    98:d9:2f:82:4d:d1:3d:38:85:4a:96:be:16:1c:c7:
                    ee:bd:46:1e:3a:3d:3a:1a:4c:0b:14:91:cd:bf:6b:
                    54:fa:da:2f:23:79:98:47:15:2e:63:85:2e:a6:44:
                    b7:30:2f:8e:b5:ec:bb:5c:a9:b5:3f:e2:0d:8d:65:
                    a8:7a:09:ba:88:20:2d:dd:5b:ac:a5:5c:fe:ea:95:
                    36:c6:09:77:9e:b0:15:31:01:c1:e1:f4:5e:40:f1:
                    f4:a5
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Attribute Descriptor: 
                Identifier: 2.5.4.3
                Syntax:
                    UnboundedDirectoryString

                Name: commonName
                Description: A general-purpose name
                Domination Rule:
                    Privilege Policy Identifier: 2.5.4.10
                    Privilege Policy Syntax:
                        Pointer:
                            Names:
                              DirName:CN = Wildboar Software
                            Hash:
                                Algorithm: sha256
                                Hash Value: 
    Signature Algorithm: sha1WithRSAEncryption
    Signature Value:
```

**Comment by t8m** [comment]
What about the empty line after the `UnboundedDirectoryString` line? Is that intentional?

**Pull request closed without merging**